### PR TITLE
Dashboard Stack

### PR DESCRIPTION
**Pull request by IceWreck**
Paging @csomh 

Link for easier reading. https://github.com/IceWreck/research/blob/master/dashboard/README.md

### CODE REVIEW

File: dashboard/README.md [hunk 1]
```diff
@@ -1,2 +1,141 @@
...
+* SCSS is a superset of CSS with variable support and better query selection. It is compiled to _readable_ CSS.
+
+### UI/UX
+
+Patternfly has two official variants.
+
+* Patternfly HTML: Its a CSS Framework. Zero inbuilt support for JavaScript. User has to take care of interactive components + animations themselves.
+* Patternfly React: Complete UI toolkit. Components are interactive by default. Inbuilt support for charts and datatables, which follow a similar design pattern.
+
+Note: They also have a slack for community support.
```
Comment (TomasTomecek) [inline dashboard/README.md:18]: ...and the core devs are working for Red Hat :)

File: dashboard/README.md [hunk 2]
```diff
@@ -1,2 +1,141 @@
...
+
+## UI
+
+### Utils
+
+* SCSS is a superset of CSS with variable support and better query selection. It is compiled to _readable_ CSS.
+
+### UI/UX
+
+Patternfly has two official variants.
```
Comment (TomasTomecek) [inline dashboard/README.md:13]: [cockpit project is using patternfly](https://github.com/cockpit-project/cockpit/tree/master/pkg) if you wanna get inspired

File: dashboard/README.md [hunk 3]
```diff
@@ -1,2 +1,141 @@
...
+Since the dashboard will fetch all data from the API, we can make it entirely client side. Using Fetch/Axios, the browser will fetch the json directly from the API, no middleman involved. Controlling the routes in the URL is possible beacuse react-router uses the HTML5 History API. After compiling the JSX to JS, we will have static HTML/CSS/JS which can be served from apache, nginx etc or even GitHub pages.
+
+#### Pros
+* Serving static files is faster than wsgi + python. 
+* One language for everything.
+* Single page web app.
+* Using React + Patternfly's extra React only features like charts and datatables is leaner than importing Patternfly HTML + jQuery + some charts library + datatables library + other plugins.
+
+
+#### Cons
```
Comment (TomasTomecek) [inline dashboard/README.md:37]: do we need to tailor the API jsons for react?

File: dashboard/README.md [hunk 4]
```diff
@@ -1,2 +1,141 @@
...
+* Number of successful out of total builds of all time. (donut chart)
+* Total number of projects (100% filled donut chart)
+* Total number of installations (100% filled donut chart)
+* Manually triggered vs automatic builds. (line chart) (toggle for weekly/monthly/yearly)
+* Top 5 most active projects. (Data List) (toggle for weekly/monthly/all-time)
+* Testing farm usage chart
+
+### Build Info Page
+
+*
```
Comment (TomasTomecek) [inline dashboard/README.md:118]: package name, project, SRPM name, a list of built RPMs, size, link to trigger (PR, release)

File: dashboard/README.md (hunk 2 above)
Comment (IceWreck) [reply dashboard/README.md:13]: Yes, that is where I first saw patternfly.

File: dashboard/README.md (hunk 3 above)
Comment (IceWreck) [reply dashboard/README.md:37]: No, we'll just fetch the json and then parse it as a regular javascipt object in the dashboard

File: dashboard/README.md (hunk 4 above)
Comment (IceWreck) [reply dashboard/README.md:118]: Yes, sorry. This PR was meant to be work in progress. I'll fill the rest of the pages later today

Comment (TomasTomecek) [reply dashboard/README.md:118]: no worries, I was really just suggesting

**Comment by csomh** [comment]
@IceWreck please also take a look at #46. We have a possibility to use an internal Prometheus and Grafana instance, so we wouldn't need to maintain them ourselves. 

I see Prometheus as working in combination with the Dashboard.

It would help us working around both the fact that visualising JSON with Grafana is not nice and that Prometheus doesn't play nicely with high-cardinality data. The dashboard would be there to give us a nice view into what Packit Service "knows", while we could use the Prometheus stack for what it is intended: monitoring.

We would count on you to work on the Dashboard and we could take care of Prometheus.

What do you think? Ping me tomorrow please and let's have a chat about this :slightly_smiling_face:

File: dashboard/README.md [hunk 5]
```diff
@@ -1,2 +1,173 @@
...
+* SRPM Name
+* Repository URL
+* Link to trigger (pull request or release)
+* Copr Build ID
+* PR ID (where applicable)
+* Project ID
+* Build Submitted, Started, Finished Time
+* List of chroots
+* Status per chroot
+* Instructions to quick enable and install that build
```
Comment (IceWreck) [inline dashboard/README.md:129]: Lets say the direct install instructions for a package are:
```
$ dnf copr enable icewreck/some-package-10
$ dnf install icewreck/some-package-10
```

Here the _'-10'_ is the GitHub PR ID according to the way packit names copr packages.

But if we do _/packit build_ manually on PR#10 then a completely new build, with a new build id is created. However the install instructions remain the same. So what should we do for the older builds of the same package? Remove direct install instructions for them?

Comment (csomh) [reply dashboard/README.md:129]: With these install instructions it's always the package from the last build is going to be installed. So it's fine for the instructions to stay the same.

I *think* packages from previous builds are still available, but one will need to specify the full version when installing. See for example [this PR](https://copr.fedorainfracloud.org/coprs/packit/beaker-project-beaker-22/builds/) which had multiple builds.

**Comment by csomh** [comment]
This might open a somewhat different direction to the current plan presented here, but I'll write it down nevertheless, and let me know @IceWreck what do you think :)

On Friday I did a few thought experiments about how I would use the Dashboard to help understanding and debugging the errors we have in the service.

One of the things that usually happens is that one of our users reports that something went wrong on on of their PRs. Given the project name and the PR ID, I'll start tracking down the issue.

If I'm lucky, the thing that went wrong ended up as an event in Sentry, where I'll have some context provided. But if this is not the case, I'll have to go to OpenShift, and hope that the pods were not redeployed since the issue happened and I'm still able to filter out some of the relevant logs to understand what Packit did for the given PR.

Or I could just craft a smart SQL query to try to extract some information about that PR from the database.

But none of the solutions above is very smart or convenient.

I would rather go to a page, presenting the list of jobs Packit did, in reverse chronological order. Here by job I mean the thing we define in under the [jobs](https://packit.dev/docs/configuration/#jobs) section in `packit.yaml`. So a PR could have multiple jobs in this list.

The list would identify jobs by project and PR id, would present the trigger (ex: `pull_request`) and the type of the job (ex: `copr_build`).

I could also see how the tasks within the jobs are doing: for example, how many chroots are building, finished building successfully or failed.

Selecting an entry would take me to a job page, with more details about the tasks (builds, tests) executed within that job. A could get the link for the log page from there, for example.

The list of jobs could be filtered using a simple search box. I could filter for jobs which run yesterday or last week. I could filter for projects, job types, triggers and so on.

The above has no charts, that's true, but a list like this would be immediately useful in helping us to understand what's going on in Packit, and as the data would be already gathered, we could easily chart this information as a follow up step.

File: dashboard/README.md [hunk 6]
```diff
@@ -1,2 +1,173 @@
...
+* Help menu, login button, settings icon, documentation buttons cannot be removed.
+* Large, publically visible buttons point to grafana documentation, which will be misleading in packit-service's dashboard.
+* We cannot control the URL scheme.
+* I haven't found a way to do something like recursive json.  Eg: (GET and display all builds info and if a user a wants it, find build_id from the all builds json and then do a GET request to fetch another json file for detailed info regarding that build_id)
+* Tests ?
+* Adapting it for our use case will be a literal hackfest with tons of plugins thrown in to make it work.
+
+
+
+## Dashboard - List of Features
```
Comment (csomh) [inline dashboard/README.md:104]: @IceWreck please update this section with what we discussed yesterday, so we can merge this :)

Comment (IceWreck) [reply dashboard/README.md:104]: Done

**Pull request merged**